b2luigi/batch/processes/htcondor.py

```diff
@@ -0,0 +1,226 @@
+
+
+class HTCondorJobStatus(enum.IntEnum):
+    """
+    See https://htcondor.readthedocs.io/en/latest/classad-attributes/job-classad-attributes.htmls
+    """
+    idle = 1
+    running = 2
+    removed = 3
+    completed = 4
+    held = 5
+
+
+_batch_job_status_cache = HTCondorJobStatusCache()
+
+
+class HTCondorProcess(BatchProcess):
+    """
+    Reference implementation of the batch process for a HTCondor batch system.
+
+    We assume that the batch system shares a file system with the submission node you
+    are currently working on (or at least the current folder and the result/log directory
+    are also available there with the same path).
+    An environment has to be set up on its own for each job. For that, a environment
+    setup script has to be provided in the ``settings.json`` file.
+    If no own python cmd is specified, the task is executed with the current ``python3``
+    available after the environment is setup.
+    General settings (may be depended on your HTCondor setup) that affect all jobs (tasks)
+    can be specified in the ``settings.json`` by adding a ``htcondor_settings`` entry.
+    Job specific settings, e.g. number of cpus or required memory can be specified by adding
+    a ``htcondor_settings`` attribute to the task. It's value has to be a dictionary containing
+    also HTCondor settings as key/value pairs.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # reassign task_cmd since the current python3 command provided by
+        # the new environment should be used by default
+        self.task_cmd, self.task_env = create_cmd_from_task(self.task, "python3")
+        self._batch_job_id = None
+
+    def get_job_status(self):
+
+        assert self._batch_job_id
+
+        try:
+            job_status = _batch_job_status_cache[self._batch_job_id]
+        except KeyError:
+            return JobStatus.aborted
+
+        if job_status in [HTCondorJobStatus.completed]:
+            return JobStatus.successful
+        elif job_status in [HTCondorJobStatus.idle, HTCondorJobStatus.running]:
+            return JobStatus.running
+        elif job_status in [HTCondorJobStatus.removed, HTCondorJobStatus.held]:
+            return JobStatus.aborted
+        else:
+            raise ValueError(f"Unknown HTCondor Job status: {job_status}")
+
+    def start_job(self):
+        submit_file_path = self._create_htcondor_submit_file()
+
+        submit_file_dir, submit_file = os.path.split(submit_file_path)
+        cmd = ["condor_submit", submit_file]
+
+        # apparently you have to be in the same directory as the submit file
+        # to be able to submit jobs with htcondor
+        cur_dir = os.getcwd()
+
+        # have to copy settings file to job working directory since b2luigi has to
+        # take the result path from it
+        shutil.copyfile("settings.json", os.path.join(submit_file_dir, "settings.json"))
```

Comment:
Just a small notice: that will not work for all cases. We can leave it as it is, I will try to come up with another solution.